### ISSUE TITLE
Performance issues with a large amount of arguments

### ISSUE THREAD

**Issue by codido** [mesonbuild/meson#1485]
Meson seems to perform poorly with a large number of compilation arguments (e.g. include directories, defines). As an example, the following can take over 17 seconds on an i7 machine and is mostly CPU bound:
```bash
#!/bin/bash

cat > meson.build <<EOF
project('example', 'c')
src = []
EOF

for i in {1..1000}
do
   touch file$i.c
   echo "add_global_arguments('-DDEFINE$i', language: 'c')" >> meson.build
   echo "src += files('file$i.c')" >> meson.build
done

echo "executable('example', src)" >> meson.build

time meson build
rm -rf file*.c build
```

**Comment by nirbheek** [comment]
This is because we use a special class for compiler arguments that maintains the appropriate order for new arguments such that they override previous ones (f.ex., includes must be added in front, and defines in the end). The implementation is not very efficient because compiler argument lists are usually small. We could probably improve that by, for instance, maintaining two lists instead of one: one list for arguments that override by appending, another for arguments that override by prepending.

The real question though is how did you notice this and is it affecting your configuration times significantly? :)

**Comment by codido** [comment]
@nirbheek I noticed this on a real project (obviously not the demo one ;)), where multiple include directories caused a noticeable slow down. I'm afraid the directory structure cannot be modified in that particular project.

By the way, __iadd__ is indeed the bottleneck here, though generating the build still takes a couple of seconds even when post.append is called blindly (just as a test).

One thing to keep in mind is that the logic is done per C file, even though the arguments are only mentioned once (whether through add_global_arguments or include_directories). I suspect that might be optimized as well, even if the actual override rules are kept as is.

**Comment by QuLogic** [comment]
This can be dropped by about a third by using an `OrderedDict`...

**Comment by jpakkane** [comment]
AFAICR we evaluate the compiler argument list for every file, even though the only thing that changes inside one target are the file names. We should probably cache that.

**Comment by QuLogic** [comment]
This looks like it's probably called for every file:
```
Mon Mar 20 18:45:08 2017    profile.stats

         16305974 function calls (16247060 primitive calls) in 19.583 seconds

   Ordered by: cumulative time
   List reduced from 1573 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     96/1    0.003    0.000   19.583   19.583 {built-in method builtins.exec}
        1    0.000    0.000   19.583   19.583 meson/meson.py:17(<module>)
        1    0.000    0.000   19.513   19.513 meson/meson.py:20(main)
        1    0.000    0.000   19.513   19.513 meson/mesonbuild/mesonmain.py:238(run)
        1    0.000    0.000   19.513   19.513 meson/mesonbuild/mesonmain.py:129(generate)
        1    0.001    0.001   19.039   19.039 meson/mesonbuild/backend/ninjabackend.py:178(generate)
        1    0.006    0.006   18.839   18.839 meson/mesonbuild/backend/ninjabackend.py:285(generate_target)
     1000    0.041    0.000   18.824    0.019 meson/mesonbuild/backend/ninjabackend.py:1805(generate_single_compile)
    17015   15.933    0.001   17.338    0.001 meson/mesonbuild/compilers.py:444(__iadd__)
     1000    0.015    0.000    8.686    0.009 meson/mesonbuild/backend/backends.py:336(generate_basic_compiler_args)
     1011    0.624    0.001    1.332    0.001 meson/mesonbuild/backend/ninjabackend.py:96(write)
  2016002    0.398    0.000    0.666    0.000 meson/mesonbuild/compilers.py:405(_can_dedup)
  2016002    0.421    0.000    0.653    0.000 meson/mesonbuild/compilers.py:430(_should_prepend)
  4065408    0.503    0.000    0.503    0.000 {method 'startswith' of 'str' objects}
  1014077    0.277    0.000    0.490    0.000 meson/mesonbuild/backend/ninjabackend.py:39(ninja_quote)
        1    0.000    0.000    0.416    0.416 meson/mesonbuild/interpreter.py:1179(__init__)
        8    0.000    0.000    0.324    0.041 /usr/lib64/python3.5/subprocess.py:1032(communicate)
  3051329    0.321    0.000    0.321    0.000 {method 'replace' of 'str' objects}
        1    0.000    0.000    0.271    0.271 meson/mesonbuild/interpreterbase.py:91(load_root_meson_file)
        1    0.000    0.000    0.269    0.269 meson/mesonbuild/mparser.py:397(parse)
```

With an `OrderedDict`:
```
Mon Mar 20 19:12:52 2017    profile.stats

         18334179 function calls (18275259 primitive calls) in 5.971 seconds

   Ordered by: cumulative time
   List reduced from 1588 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     97/1    0.003    0.000    5.971    5.971 {built-in method builtins.exec}
        1    0.000    0.000    5.971    5.971 meson/meson.py:17(<module>)
        1    0.000    0.000    5.841    5.841 meson/meson.py:20(main)
        1    0.000    0.000    5.840    5.840 meson/mesonbuild/mesonmain.py:238(run)
        1    0.000    0.000    5.840    5.840 meson/mesonbuild/mesonmain.py:129(generate)
        1    0.001    0.001    5.346    5.346 meson/mesonbuild/backend/ninjabackend.py:178(generate)
        1    0.006    0.006    5.142    5.142 meson/mesonbuild/backend/ninjabackend.py:285(generate_target)
     1000    0.069    0.000    5.128    0.005 meson/mesonbuild/backend/ninjabackend.py:1805(generate_single_compile)
    17015    1.902    0.000    3.616    0.000 meson/mesonbuild/compilers.py:446(__iadd__)
     1000    0.040    0.000    1.884    0.002 meson/mesonbuild/backend/backends.py:336(generate_basic_compiler_args)
     1011    0.640    0.001    1.305    0.001 meson/mesonbuild/backend/ninjabackend.py:96(write)
  2016002    0.382    0.000    0.661    0.000 meson/mesonbuild/compilers.py:406(_can_dedup)
  2016002    0.372    0.000    0.603    0.000 meson/mesonbuild/compilers.py:431(_should_prepend)
  4065423    0.513    0.000    0.513    0.000 {method 'startswith' of 'str' objects}
  1014077    0.278    0.000    0.491    0.000 meson/mesonbuild/backend/ninjabackend.py:39(ninja_quote)
  4010000    0.446    0.000    0.446    0.000 {method 'pop' of 'collections.OrderedDict' objects}
        1    0.000    0.000    0.436    0.436 meson/mesonbuild/interpreter.py:1184(__init__)
        8    0.000    0.000    0.348    0.043 /usr/lib64/python3.5/subprocess.py:1032(communicate)
  3051330    0.321    0.000    0.321    0.000 {method 'replace' of 'str' objects}
        1    0.000    0.000    0.271    0.271 meson/mesonbuild/interpreterbase.py:91(load_root_meson_file)
```

Caching would probably cut most of that out I guess.

**Comment by QuLogic** [comment]
`generate_single_compile` called 1000 times (once per file) and calls `generate_basic_compiler_args` 1000 times as well:
![1000call](https://cloud.githubusercontent.com/assets/302469/24126800/6ce121bc-0da6-11e7-8ddf-d04ab5cba593.png)

**Comment by QuLogic** [comment]
Switching `pre.pop(item, None)` to `if item in pre: del pre[item]` seems to cut it down to ~5 seconds and then adding a cache to `generate_basic_compiler_args` cuts it down to ~3.5s. There's a mix of global and target-specific arguments in that function, so I'm not totally sure I've done that right.

**Comment by nirbheek** [comment]
In addition to creating all these caches, I wonder if it makes sense to create per-target rules in the `build.ninja` file. Those are for exactly this purpose, and it would also:

1. Speed up writing of the Ninja file
2. Reduce the size of the `build.ninja` file enormously
3. Improve Ninja's parsing of the `build.ninja` file

We could in fact make things even nicer by having a rule like:

```ninja
incpaths = -Ifoo -Ibar
libpaths = -Lfdir -Lbdir
args = -e -v -e -r -y -t -h -i -n -g -else
rule sometarget
  command = gcc $incpaths $libpaths $in $args -o $out 
```

(or similar).

**Comment by QuLogic** [comment]
So this can be closed by #1486.
